### ISSUE TITLE
Show untracked files in magit status

### ISSUE THREAD

**Issue by roblabla** [kahole/edamagit#33]
Emacs magit has an option to allow showing a tree of untracked files in the magit status:

![](https://opensource.com/sites/default/files/uploads/magit_status.png)

It'd be nice if edamagit could have an option to show those as well.

**Comment by kahole** [comment]
It should do that already. What platform are you on?

**Comment by roblabla** [comment]
It's on windows. `$` is empty. Here's a screenshot showing the issue: 
![image](https://user-images.githubusercontent.com/1069318/82364161-30ac3300-9a0f-11ea-9336-2454847b1027.png)

**Comment by kahole** [comment]
Whats the output of
```git ls-files --others --exclude-standard --directory```
when run in from the repository root?

**Comment by roblabla** [comment]
```
E:\dev\projects\ubuntu-installer> git ls-files --others --exclude-standard --directory
bindings/target/debug/build/
bindings/target/debug/deps/
bindings/target/debug/examples/
bindings/target/debug/incremental/winrt-i3qifj4o32hd/s-fniqinkuht-sqp1ij-working/
src/wizard.rs
ui-research/
```

**Comment by kahole** [comment]
Is there anything in the 'Developer Tools -> Console' of VSCode, when you open Magit status?

**Comment by roblabla** [comment]
TLDR: I managed to track down how to reproduce the issue. It's caused by the `git.untrackedChanges: "separate"` setting. It appears to be a defect in the VSCode git API, I'm not sure we can work around it.

-------

So I ran the code under the vscode debugger, and managed to get a clearer view of what's going on. https://github.com/kahole/edamagit/blob/develop/src/commands/statusCommands.ts#L77 the `git ls-files` command is only run if `repository.state.workingTreeChanges` contains UNTRACKED changes. However, for some reason, `repository.state` does not contain any of the UNTRACKED changes on my computer. It appears to only contain MODIFIED changes.

I dug a bit into how vscode's status function is implemented. It calls `git status -z -u` (see https://github.com/microsoft/vscode/blob/master/extensions/git/src/git.ts#L1726). This output then gets parsed by Repository.status. The untracked file shows up as an `??` prefix: https://github.com/microsoft/vscode/blob/master/extensions/git/src/repository.ts#L1564

Here's the thing: I enabled the `git.untrackedChanges: "separate"` option (which separates the untracked files from the rest of the changes in the vscode "Source Control" view). This option also affects the Git API! When enabling this option, the untracked file changes don't go in `workingTreeChanges` like the rest, but in a untrackedChanges list.

I'm not entirely sure it's possible to actually access that list through the public API... This feels like a defect in vscode. Should we open an issue on the vscode tracker?

**Comment by roblabla** [comment]
Also note that it's possible to do `git.untrackedChanges: "hide"` in which case the git extension will completely ignore the untracked changes, making magit unable to see them at all! This makes me doubly think this is a defect in vscode. The `git.status()` API shouldn't be impacted by UI settings...

**Comment by kahole** [comment]
Interesting. I agree that it's strange that it affects the API, or that there isn't a separate field for the untracked changes in the API always, or just when this setting is in effect.
I see in the docs of `git.untrackedChanges: "separate"` that it can affect the behaviour of commands and are excluded from some 'actions'..

An easy fix is of course to always run the `git ls-files` without checking for 'untracked' changes beforehand. I included the check for efficiency.

I see that the untracked files are detectable in the inner _repository state of the the git extension in the property `untrackedGroup`, but I'm not sure I wanna add even more workarounds of the API.

In fact a long term goal is to move away from using the API at all, as I don't find it flexible or efficient enough for the stuff I wanna do. This will take some work, though. Today the extension is a hybrid of vscode-commands, git API calls, and raw git commands.

**Comment by kahole** [comment]
Closing as I don't think there's anything more to do here just now. Relying less on vscodes git API would fix this, which is sort of planned.
